java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java

```diff
@@ -0,0 +1,166 @@
+import org.sonar.java.locks.LockState;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.CheckForNull;
+import java.util.Map;
+import java.util.Set;
+
+public class ExecutionState {
+
+  private final IssuableSubscriptionVisitor check;
+  ExecutionState parent;
+  private Map<Symbol, Value> valuesBySymbol = Maps.newHashMap();
+
+  public ExecutionState(ExecutionState executionState) {
+    this.parent = executionState;
+    this.check = executionState.check;
+  }
+
+  public ExecutionState(IssuableSubscriptionVisitor check) {
+    this.check = check;
+  }
+
+  public ExecutionState merge(ExecutionState executionState) {
+    for (Map.Entry<Symbol, Value> entry : executionState.valuesBySymbol.entrySet()) {
+      Symbol symbol = entry.getKey();
+      Value currentValue = getValue(symbol);
+      Value valueToMerge = entry.getValue();
+      if (currentValue != null) {
+        currentValue.state = currentValue.state.merge(valueToMerge.state);
+        valuesBySymbol.put(symbol, currentValue);
+      } else {
+        // possible way to solve the problem of variable defined in outer Execution state. lockedOccurencesBySymbol.put(symbol,
+        // occurenceToMerge);
+        if (valueToMerge.shouldRaiseIssue()) {
+          insertIssue(valueToMerge.treeNode);
+        }
+      }
+    }
+    return this;
+  }
+
+  public ExecutionState overrideBy(ExecutionState executionState) {
+    for (Map.Entry<Symbol, Value> entry : executionState.valuesBySymbol.entrySet()) {
+      Symbol symbol = entry.getKey();
+      Value value = entry.getValue();
+      if (getValue(symbol) != null) {
+        markValueAs(symbol, value.state);
+      } else {
+        valuesBySymbol.put(symbol, value);
+      }
+    }
+    return this;
+  }
+
+  public void newValueForSymbol(Symbol symbol, Tree tree, State state) {
+    Value knownValue = getValue(symbol);
+    if (knownValue != null) {
+      if (knownValue.shouldRaiseIssue()) {
+        insertIssue(knownValue.treeNode);
+      }
+    } else {
+      // no known occurence, means its a field or a method param.
+      createValueInTopExecutionState(symbol, tree, LockState.NULL);
+    }
+    valuesBySymbol.put(symbol, new Value(tree, state));
+  }
+
+  private void createValueInTopExecutionState(Symbol symbol, Tree tree, State state) {
+    ExecutionState top = this;
+    while (top.parent != null) {
+      top = top.parent;
+    }
+    top.valuesBySymbol.put(symbol, new Value(tree, state));
+  }
+
+  private void insertIssue(Tree treeNode) {
+    check.addIssue(treeNode, "");
+  }
+
+  public void markValueAs(Symbol symbol, State state) {
+    Value value = getValue(symbol);
+    if (value != null) {
+      valuesBySymbol.put(symbol, new Value(value.treeNode, state));
+    }
+  }
+
+  public ExecutionState restoreParent() {
+    if (parent != null) {
+      // insertIssues();
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)This block of commented-out lines of code should be removed.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)